lgc/patch/PatchBoundsCheckMemory.cpp

```diff
@@ -0,0 +1,195 @@
+ ***********************************************************************************************************************
+ */
+#include "PatchBoundsCheckMemory.h"
+#include "lgc/state/IntrinsDefs.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "lgc-bounds-check-memory"
+
+using namespace llvm;
+using namespace lgc;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Initializes static members.
+char PatchBoundsCheckMemory::ID = 0;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of bounds check memory operations.
+ModulePass *createPatchBoundsCheckMemory() {
+  return new PatchBoundsCheckMemory();
+}
+
+static Value *getTypeZero(Type *ty) {
+  if (ty->isIntegerTy()) {
+    return ConstantInt::get(ty, 0);
+  } else if (ty->isFloatTy() || ty->isHalfTy() || ty->isDoubleTy() || ty->isFP128Ty() || ty->isBFloatTy()) {
+    return ConstantFP::get(ty, 0.0);
+  } else {
+    return ConstantAggregateZero::get(ty);
+  }
+}
+
+// =====================================================================================================================
+PatchBoundsCheckMemory::PatchBoundsCheckMemory() : Patch(ID) {
+}
+
+// =====================================================================================================================
+// Executes this pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+bool PatchBoundsCheckMemory::runOnModule(Module &module) {
+  LLVM_DEBUG(dbgs() << "Run the pass llpc-bounds-check-memory\n");
+
+  m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(&module);
+
+  if (!m_pipelineState->getOptions().enableScratchBoundsCheck)
+    return false;
+
+  m_context = &module.getContext();
+  m_builder.reset(new IRBuilder<>(*m_context));
+
+  visit(module);
+
+  bool changed = false;
+  for (auto &inst : m_getElementPtrInsts)
+    changed |= addBoundsCheck(inst);
+  m_getElementPtrInsts.clear();
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Visits "getelementptr" instruction.
+//
+// @param getElemPtr : "GetElementPtr" instruction
+void PatchBoundsCheckMemory::visitGetElementPtrInst(GetElementPtrInst &getElemPtr) {
+  auto pointerTy = getElemPtr.getPointerOperand()->getType();
+  if (pointerTy->getPointerAddressSpace() != ADDR_SPACE_PRIVATE)
+    return;
+
+  // Check for dynamic indices in the instruction
+  llvm::SmallVector<std::pair<Value *, uint64_t>, 1> dynIndices;
+  std::vector<Value *> indices;
+  for (unsigned i = 1, operandCount = getElemPtr.getNumOperands(); i < operandCount; ++i) {
+    auto index = getElemPtr.getOperand(i);
+    if (!isa<Constant>(index)) {
```

Comment:
can you flip this condition to avoid nesting the code so much?
i.e., `if (isa<Constant>(index)) continue;`

Reply:
The `break` at the end of the if should not have been there, so I can’t do that without duplicating the `indices.push_back`.